Auto-detect whether Benchmark should produce colorized output

nickhutchinson opened the pull request:
Attached is a patch that makes Benchmark only produce colorized output
if stdout is a TTY. (This can be overridden with --color-print=true.)

The logic for deciding whether to use colorized output was lifted from
GTest: <https://code.google.com/p/googletest/source/browse/trunk/src/gtest.cc#2872>.

I haven't got a Windows system to test my changes on, but I see from #7
that there's more work to be done here anyway, so hopefully that's okay!

EricWF commented on src/posix.h:
```diff
@@ -0,0 +1,57 @@
+#endif
+
+namespace benchmark {
+namespace posix {
```
I know the top level namespace `posix` is reserved for the implementation in C++. I would prefer not using a nested one either. Could you come up with another name?

nickhutchinson replied:
My model here was Google Test's [internal/gtest-port.h](https://code.google.com/p/googletest/source/browse/trunk/include/gtest/internal/gtest-port.h#2239). I'll have a ponder about an alternate namespace name. Would you be open to ditching the nested namespace altogether?

EricWF replied:
Not opposed towards dropping the namespace.

EricWF commented on AUTHORS:
```diff
@@ -19,6 +19,7 @@ Google Inc.
 JianXiong Zhou <zhoujianxiong2@gmail.com>
 Lei Xu <eddyxu@gmail.com>
 Matt Clarkson <mattyclarkson@gmail.com>
+Nick Hutchinson <nshutchinson@gmail.com>
```
This doesn't really matter, but what is the difference between "authors" and "contributors"?

dominichamon replied:
AUTHORS are people who own the copyright, ie, if a person signs the CLA on behalf of a company, that company should be in here.

CONTRIBUTORS are people who contribute but do not have copyright necessarily.

Everyone who contributes should be in CONTRIBUTORS. Everyone who claims copyright should be in AUTHORS.

EricWF commented on src/benchmark.cc:
```diff
@@ -877,8 +881,8 @@ void ParseCommandLineFlags(int* argc, const char** argv) {
                         &FLAGS_benchmark_format) ||
-        ParseBoolFlag(argv[i], "color_print",
-                       &FLAGS_color_print) ||
+        ParseStringFlag(argv[i], "color_print",
```
I think the value of this switch should be verified below.

nickhutchinson replied:
How lenient should the check be? ParseBoolFlag() is quite generous with what it accepts, as is Google Test's --gtest_color option. 

I've changed it to assert that the user passes a non-empty string. Is that enough?

EricWF commented on src/colorprint.cc:
```diff
@@ -73,13 +74,50 @@ PlatformColorCode GetPlatformColorCode(LogColor color) {
+
+// Returns true iff Benchmark should use colors in the output.
+bool ShouldUseColor(bool stdout_is_tty) {
+  const char* const color_print = FLAG(color_print).c_str();
```
It seems like it would be a lot easier to compare a std::string than a C string.

```c++
std::string const& color_print = FLAG(color_print);
const bool term_supports_color =
   term == "xterm" ||
   term == "foo" ||
  ...;
```

EricWF commented:
Seems like a reasonable patch. Is there a reason to perfer using C strings in the function interfaces as opposed to std::string? If there isn't then I would prefer to use std::string.

dominichamon commented on src/benchmark.cc:
```diff
@@ -858,7 +862,7 @@ void PrintUsageAndExit() {
           "          [--benchmark_repetitions=<num_repetitions>]\n"
           "          [--benchmark_format=<tabular|json|csv>]\n"
-          "          [--color_print={true|false}]\n"
+          "          [--color_print={auto|true|false}]\n"
```
the usage says 'true' and 'false' while the help says 'yes' and 'no'.

nickhutchinson replied:
Fixed

dominichamon commented on src/benchmark.cc:
```diff
@@ -67,7 +67,11 @@ DEFINE_string(benchmark_format, "tabular",
               "'tabular', 'json', or 'csv'.");
 
-DEFINE_bool(color_print, true, "Enables colorized logging.");
+DEFINE_string(color_print, "auto",
```
a boolean might be better with default 'true' being overridden if the TERM doesn't support it.

nickhutchinson replied:
I think the three states are necessary: it's sometimes useful to pipe to some command that *does* support colorized output. In this case, the `isatty()` test would fail, so you do need a way to forcibly enable colors.

dominichamon commented on src/benchmark.cc:
```diff
@@ -877,14 +882,17 @@ void ParseCommandLineFlags(int* argc, const char** argv) {
       --(*argc);
       --i;
-    } else if (IsFlag(argv[i], "help")) {
+    } else if (IsFlag(argv[i], "help") || HasBenchmarkFlagPrefix(argv[i])) {
```
!HasBenchmarkFlagPrefix?

also, redundant given how we parse the strings above.

nickhutchinson replied:
Not quite -- without this check, you could pass `--benchmark_format` or similar with no associated value, and no error will be reported.

dominichamon replied:
$ test/benchmark_test --benchmark_format=foo
benchmark [--benchmark_list_tests={true|false}]
          [--benchmark_filter=<regex>]
          [--benchmark_min_time=<min_time>]
          [--benchmark_repetitions=<num_repetitions>]
          [--benchmark_format=<tabular|json|csv>]
          [--color_print={true|false}]
          [--v=<verbosity>]

this is working as intended. the checks for specific args come next.

dominichamon commented on src/commandlineflags.cc:
```diff
@@ -216,4 +209,34 @@ bool ParseStringFlag(const char* str, const char* flag, std::string* value) {
+          SkipPrefix("benchmark-", &str));
+}
+
+bool IsTruthyFlagValue(const char* value)
```
this should be taking a string.

nickhutchinson replied:
Is there a reason this one should take a std::string when the others in the header take a c string?

dominichamon replied:
I think they should all be taking std::string. @EricWF might have a preference though as some headers should be C-ish only.

EricWF replied:
As long as the std::string doesn't cross the DSO boundary. Internally I agree we should use std::string.

dominichamon commented on src/commandlineflags.cc:
```diff
@@ -216,4 +209,34 @@ bool ParseStringFlag(const char* str, const char* flag, std::string* value) {
+
+// If |*pstr| begins with |prefix|, advance *pstr so that it points
+// just past the prefix, and return true. Otherwise, return false.
+static bool SkipPrefix(const char* prefix, const char** pstr) {
```
neither of these functions are useful.

nickhutchinson replied:
Could you elaborate?

dominichamon replied:
we don't need HasBenchmarkFlagPrefix (see earlier comment) and SkipPrefix is only used by it.

For colored output, this is not necessary.

dominichamon commented:
@nickhutchinson are you still working on this?

EricWF commented:
@nickhutchinson Are you interested in continuing with this?

nickhutchinson commented:
Sorry for the delay, I’ll try fix this up this week.

> On 29 Aug 2016, at 06:21, Eric <notifications@github.com> wrote:
> 
> @nickhutchinson <https://github.com/nickhutchinson> Are you interested in continuing with this?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/google/benchmark/pull/126#issuecomment-243034408>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAuH7KjeTf9hqmHHi02JX-Sc_iWk-dL4ks5qkmx2gaJpZM4EzKUw>.
>

nickhutchinson commented:
Finally got around to rebasing this on master -- thanks for your patience. I believe I've addressed the review comments, in that the new patch:
- uses std::string instead of const char*
- removes error-checking of CLI arguments -- as pointed out, this isn't relevant to colorised output.

I've tested on Win10/MSVC2015 and OS X 10.11/Clang 3.8.

EricWF commented on src/posix.h:
```diff
@@ -0,0 +1,45 @@
+  return (env != nullptr && env[0] != '\0') ? env : nullptr;
+#else
+  return getenv(name);
+#endif // BENCHMARK_OS_WINDOWS
```
The comment is incorrect.

EricWF commented on src/posix.h:
```diff
@@ -0,0 +1,45 @@
+#endif // BENCHMARK_OS_WINDOWS
+}
+
+inline int IsATTY(int fd) {
```
Let's make this idiomatic C++ and change the return type to `bool`

EricWF replied:
Actually since `IsATTY` and `FileNo` seem like unneeded abstractions since they only have one caller, `IsColorTerminal`. I would remove these and just call `isatty` and `fileno` directly from there.

EricWF commented on src/posix.h:
```diff
@@ -0,0 +1,45 @@
+#ifndef BENCHMARK_POSIX_H_
```
I'm concerned about naming a header `posix.h` since it's a very generic name. Maybe `posix_support.h` or `benchmark_posix_support.h`

EricWF commented on src/commandlineflags.cc:
```diff
@@ -175,7 +167,7 @@ bool ParseBoolFlag(const char* str, const char* flag, bool* value) {
 
   // Converts the string value to a bool.
-  *value = !(*value_str == '0' || *value_str == 'f' || *value_str == 'F');
+  *value = value_str[0] == '\0' || IsTruthyFlagValue(value_str);
```
This now returns `true` for the empty string. Simply calling `IsTruthyFlagValue` should give you the right answer.

EricWF commented on src/benchmark.cc:
```diff
@@ -1123,8 +1135,11 @@ void ParseCommandLineFlags(int* argc, char** argv) {
-                       &FLAGS_color_print) ||
+        ParseStringFlag(argv[i], "benchmark_color",
+                        &FLAGS_benchmark_color) ||
+        // Legacy flag for benchmark_color.
```
Add a `TODO` or `FIXME` tag to this comment so we can grep for it.

EricWF commented on src/colorprint.h:
```diff
@@ -23,6 +24,10 @@ std::string FormatString(const char* msg, ...);
 
+// Returns true if the stream identified by 'f' appears to be a terminal that
+// supports colored output, false otherwise.
+bool IsColorTerminal(FILE* f);
```
We only really support color output to `stdout`, not arbitrary files. Therefore I think this should be `StdoutIsColorTerminal()` or similar.

EricWF commented:
Thanks again for this patch!

PS. I really liked your command line argument checking. If you want to submit that as a separate patch it would be  very welcome.

nickhutchinson commented:
Addressed code review comments.

- rebase on master
- make IsTruthyFlagValue() return true if given the empty string.
- remove posix.h header, folded the functions into IsColorTerminal() (which no longer accepts any arguments)
- remove GetEnv() shim, update callers to use stdlib's getenv() directly.
- add FIXME to --color_print comment.
